aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -238,21 +244,44 @@ SparseCsrTensor new_csr_tensor(const TensorOptions& options) {
       DispatchKeySet(dispatch_key), layout, options.dtype());
 }
 
-Tensor _sparse_csr_tensor_unsafe(const Tensor& crow_indices, const Tensor& col_indices,
-    const Tensor& values,
-    IntArrayRef size,
-    c10::optional<ScalarType> dtype,
-    c10::optional<Layout> layout,
-    c10::optional<Device> device,
-    c10::optional<bool> pin_memory) {
-
-  TensorOptions options = TensorOptions().dtype(dtype).layout(layout).device(device).pinned_memory(pin_memory);
-
-  SparseCsrTensor self = new_csr_tensor(options);
-  get_sparse_csr_impl(self)->set_member_tensors(crow_indices, col_indices, values, size);
+template <c10::Layout required_layout>
+Tensor _sparse_compressed_tensor_unsafe_template(const Tensor& compressed_indices,
+                                                 const Tensor& plain_indices,
+                                                 const Tensor& values,
+                                                 IntArrayRef size,
+                                                 c10::optional<ScalarType> dtype,
+                                                 c10::optional<Layout> layout,
+                                                 c10::optional<Device> device,
+                                                 c10::optional<bool> pin_memory) {
+
+  Layout layout_ = layout.value_or(required_layout);
```

Comment:
I think it's fine to hardcode this code for each layout if it saves us from creating a global "Unspecified" layout. You can still abstract away the code on lines 264 to 266.

Reply:
The `Unspecified` layout was introduced due to the native function
```
_sparse_compressed_tensor_unsafe(Tensor compressed_indices, Tensor plain_indices, Tensor values, int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
that requires that `layout` is specified by the user (although the signature species it as optional which is due to other peculiarities in tensor factory functions) and hence it cannot be hardcoded for `_sparse_compressed_tensor_unsafe`.

Another reason is that the code for other factory functions is considerably simpler when using `template <Layout required_layout>` approach when compared to the alternative using `c10::optional<Layout> required_layout` argument approach.

Reply:
It can.

The blocker for me is
https://github.com/pytorch/pytorch/blob/1b5c205697067c49028396eb7aa301650c14d7c6/torch/csrc/utils/tensor_new.cpp#L679
that works only if `required_layout` is a template parameter. Say, if the `required_layout` would be an argument to `_sparse_compressed_tensor_unsafe_ctor_template`, the above enum trick would not be allowed and the respective code would need to blow up as `_sparse_compressed_tensor_unsafe_ctor` and `_sparse_csr_tensor_unsafe_ctor` Python signatures are different: `_sparse_compressed_tensor_unsafe_ctor` has `layout` parameter but `_sparse_csr_tensor_unsafe_ctor` doesn't have it. By a "blow up" I mean that `_sparse_compressed_tensor_unsafe_ctor` and `_sparse_csr_tensor_unsafe_ctor` would have basically different implementations while in the current PR, these share the same implementation (defined by the template).

Reply:
I think it's fine for those to be different implementations for now instead of introducing this new concept. We can clean that up later on.